File: src/main/java/aurocosh/divinefavor/common/potions/blends/PotionHuntersAura.java
package aurocosh.divinefavor.common.potions.blends;

import aurocosh.divinefavor.common.config.common.ConfigPresence;
import aurocosh.divinefavor.common.custom_data.player.PlayerData;
import aurocosh.divinefavor.common.custom_data.player.data.aura.hunters.HuntersAuraData;
import aurocosh.divinefavor.common.potions.base.effect.ModEffect;
import aurocosh.divinefavor.common.potions.base.potion.ModPotion;
import aurocosh.divinefavor.common.potions.common.ModBlendEffects;
import aurocosh.divinefavor.common.potions.common.ModBlessings;
import aurocosh.divinefavor.common.spirit.ModSpirits;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.monster.IMob;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.util.DamageSource;
import net.minecraft.util.EntityDamageSourceIndirect;
import net.minecraftforge.event.entity.living.LivingDeathEvent;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.eventhandler.EventPriority;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;

@Mod.EventBusSubscriber
public class PotionHuntersAura extends ModPotion {
    public PotionHuntersAura() {
        super("hunters_aura", true, 0x7FB8A4);
    }

    @Override
    protected void onPotionAdded(EntityLivingBase livingBase) {
        super.onPotionAdded(livingBase);
        if (!(livingBase instanceof EntityPlayer))
            return;
        EntityPlayer player = (EntityPlayer) livingBase;
        HuntersAuraData auraData = PlayerData.get(player).getHuntersAuraData();
        auraData.reset();
    }

    @SubscribeEvent(priority = EventPriority.LOWEST)
    public static void onMobDeath(LivingDeathEvent event) {
        if (!ModSpirits.arbow.isActive())
            return;
        DamageSource source = event.getSource();
        if (!(source instanceof EntityDamageSourceIndirect))
            return;
        EntityDamageSourceIndirect sourceIndirect = (EntityDamageSourceIndirect) source;
        if(!sourceIndirect.damageType.equals("arrow"))
            return;
        Entity attacker = source.getTrueSource();
        if (!(attacker instanceof EntityPlayer))
            return;
        EntityPlayer player = (EntityPlayer) attacker;
        if (!player.isPotionActive(ModBlendEffects.hunters_aura))
            return;
        Entity mob = event.getEntity();
        if (!(mob instanceof IMob))
            return;
        HuntersAuraData huntersAuraData = PlayerData.get(player).getHuntersAuraData();
        if (huntersAuraData.tryLuck()) {
            player.removePotionEffect(ModBlendEffects.hunters_aura);
            player.addPotionEffect(new ModEffect(ModBlessings.predatory_presence, ConfigPresence.predatoryPresence.duration));
        }
    }

    @Override
    public boolean isReady(int duration, int amplifier) {
        return true;
    }
}

File: src/main/java/aurocosh/divinefavor/common/potions/blends/PotionHuntersAura.kt
package aurocosh.divinefavor.common.potions.blends

import aurocosh.divinefavor.common.config.common.ConfigPresence
import aurocosh.divinefavor.common.constants.ConstMisc
import aurocosh.divinefavor.common.custom_data.player.PlayerData
import aurocosh.divinefavor.common.potions.base.effect.ModEffect
import aurocosh.divinefavor.common.potions.base.potion.ModPotion
import aurocosh.divinefavor.common.potions.common.ModBlendEffects
import aurocosh.divinefavor.common.potions.common.ModBlessings
import aurocosh.divinefavor.common.spirit.ModSpirits
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.monster.IMob
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.util.EntityDamageSourceIndirect
import net.minecraftforge.event.entity.living.LivingDeathEvent
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.common.eventhandler.EventPriority
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent

@Mod.EventBusSubscriber(modid = ConstMisc.MOD_ID)
class PotionHuntersAura : ModPotion("hunters_aura", true, 0x7FB8A4) {

    override fun onPotionAdded(livingBase: EntityLivingBase) {
        super.onPotionAdded(livingBase)
        if (livingBase !is EntityPlayer)
            return
        val auraData = PlayerData.get(livingBase).huntersAuraData
        auraData.reset()
    }

    override fun isReady(duration: Int, amplifier: Int): Boolean {
        return true
    }

    companion object {
        @SubscribeEvent(priority = EventPriority.LOWEST)
        fun onMobDeath(event: LivingDeathEvent) {
            if (!ModSpirits.arbow.isActive)
                return
            val source = event.source as? EntityDamageSourceIndirect ?: return
            if (source.damageType != "arrow")
                return
            val attacker = source.trueSource as? EntityPlayer ?: return
            if (!attacker.isPotionActive(ModBlendEffects.hunters_aura))
                return
            val mob = event.entity
            if (mob !is IMob)
                return
            val huntersAuraData = PlayerData.get(attacker).huntersAuraData
            if (huntersAuraData.tryLuck()) {
                attacker.removePotionEffect(ModBlendEffects.hunters_aura)
                attacker.addPotionEffect(ModEffect(ModBlessings.predatory_presence, ConfigPresence.predatoryPresence.duration))
            }
        }
    }
}
